editor/src/components/canvas/dom-walker.ts

```diff
@@ -94,24 +101,111 @@ function isScene(node: Node): node is HTMLElement {
   )
 }
 
+function findParentScene(target: HTMLElement): string | null {
+  const sceneID = getDOMAttribute(target, 'data-utopia-scene-id')
+  if (sceneID != null) {
+    return sceneID
+  } else {
+    if (target.parentElement != null) {
+      return findParentScene(target.parentElement)
+    } else {
+      return null
+    }
+  }
+}
+
+const LogDomWalkerPerformance = !PRODUCTION_ENV && typeof window.performance.mark === 'function'
+
+function lazyValue<T>(getter: () => T) {
+  let alreadyResolved = false
+  let resolvedValue: T
+  return () => {
+    if (!alreadyResolved) {
+      resolvedValue = getter()
+      alreadyResolved = true
+    }
+    return resolvedValue
+  }
+}
+
 export function useDomWalker(props: CanvasContainerProps): React.Ref<HTMLDivElement> {
   const containerRef = React.useRef<HTMLDivElement>(null)
+  const rootMetadataInStateRef = useRefEditorState((store) => store.editor.domMetadataKILLME)
+  const invalidatedSceneIDsRef = React.useRef<Array<string>>([])
   const selectedViews = useEditorState(
     (store) => store.editor.selectedViews,
     'useDomWalker selectedViews',
   )
 
+  const resizeObserver = React.useMemo(() => {
+    if (ObserversAvailable) {
+      return new ResizeObserver((entries: any) => {
+        for (let entry of entries) {
+          const sceneID = findParentScene(entry.target)
+          if (
+            sceneID != null &&
+            invalidatedSceneIDsRef.current != null &&
+            invalidatedSceneIDsRef.current.indexOf(sceneID) < 0
+          ) {
+            invalidatedSceneIDsRef.current.push(sceneID)
+          }
+        }
+      })
+    } else {
+      return null
+    }
+  }, [invalidatedSceneIDsRef])
+
+  const mutationObserver = React.useMemo(() => {
+    if (ObserversAvailable) {
+      return new (window as any).MutationObserver((mutations: MutationRecord[]) => {
+        for (let mutation of mutations) {
+          if (
+            mutation.attributeName === 'style' ||
+            mutation.addedNodes.length > 0 ||
+            mutation.removedNodes.length > 0
+          ) {
+            if (mutation.target instanceof HTMLElement) {
+              const sceneID = findParentScene(mutation.target)
+              if (
+                sceneID != null &&
+                invalidatedSceneIDsRef.current != null &&
+                invalidatedSceneIDsRef.current.indexOf(sceneID) < 0
+              ) {
+                invalidatedSceneIDsRef.current.push(sceneID)
+              }
+            }
+          }
+        }
+      })
+    } else {
+      return null
+    }
+  }, [invalidatedSceneIDsRef])
```

Comment:
I think we should have these as `[]` empty arrays, and even leave a comment here saying that this should only evaluate once. because we are missing the teardown functions so if someone accidentally starts making multiple resize observers, we are in trouble!

Reply:
btw AFAIK the useMemo lint rule will not be angry if you omit refs because it knows that refs are referentially stable, so including them doesn't matter